proxy/HostStatus.cc

```diff
@@ -0,0 +1,89 @@
+/** @file
+
+  Implementation of Host Proxy routing
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+#include "HostStatus.h"
+
+HostStatus *HostStatus::p_instance = nullptr;
+
+HostStatus::HostStatus() : hosts_statuses(ink_hash_table_create(InkHashTableKeyType_String))
+{
+  ink_mutex_init(&hosts_statuses_mutex);
+}
+
+HostStatus::~HostStatus()
+{
+  InkHashTableIteratorState ht_iter;
+  InkHashTableEntry *ht_entry = nullptr;
+  ht_entry                    = ink_hash_table_iterator_first(hosts_statuses, &ht_iter);
+
+  while (ht_entry != nullptr) {
+    char *value = static_cast<char *>(ink_hash_table_entry_value(hosts_statuses, ht_entry));
+    ats_free(value);
+    ht_entry = ink_hash_table_iterator_next(hosts_statuses, &ht_iter);
+  }
+  ink_hash_table_destroy(hosts_statuses);
+
+  ink_mutex_destroy(&hosts_statuses_mutex);
+}
+
+HostStatus *
+HostStatus::instance()
+{
+  return p_instance;
+}
+
+HostStatus *
+HostStatus::init()
+{
+  if (!p_instance) {
+    p_instance = new HostStatus();
+  }
+  return p_instance;
+}
+
+void
+HostStatus::setHostStatus(const char *key, const HostStatus_t status)
+{
+  HostStatus_t *_status;
+
+  ink_mutex_acquire(&hosts_statuses_mutex);
+  if (ink_hash_table_lookup(hosts_statuses, key, (void **)&_status)) {
+    *_status = status;
+  } else {
+    Debug("hosts_statuses", "In HostConfigParams::setHostStatus(): key: %s, status: %d", key, status);
+    _status  = static_cast<HostStatus_t *>(ats_malloc(sizeof(HostStatus_t)));
```

Comment:
Any reason for using `ats_malloc` instead of `new`? In this case the constructor will not be invoked.

Reply:
Actually, reading more, why allocate anything? The `HostStatus_t` will fit in a `void *` so just store it directly.

Reply:
@SolidWallOfCode not sure how to eliminate the allocation.  Could you show me what you had in mind to use the InkHashTableEntry pointer to store the HostStatus_t value?

Reply:
Example code
```
void
HostStatus::setHostStatus(const char *key, HostStatus_t status)
{
  ink_mutex_acquire(&hosts_statuses_mutex);
  ink_hash_table_insert(hosts_statuses, key, reinterpret_cast<void*>(status));
  ink_mutex_release(&hosts_statuses_mutex);
}

HostStatus_t
HostStatus::getHostStatus(const char *key)
{
  HostStatus_t status;
  if (ink_hash_table_lookup(hosts_statuses, key, reinterpret_cast<void **>(&status))) {
    Debug("hosts_statuses", "In HostConfigParams::getHostStatus(): key: %s, status: %d", key, status);
    return status;
  }
  return HOST_STATUS_INIT;
}
```